// @grpc/grpc-js-xds/build/src/http-filter/router-filter.js
"use strict";
/*
 * Copyright 2021 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setup = void 0;
const grpc_js_1 = require("@grpc/grpc-js");
const http_filter_1 = require("../http-filter");
var BaseFilter = grpc_js_1.experimental.BaseFilter;
class RouterFilter extends BaseFilter {
}
class RouterFilterFactory {
    constructor(config, overrideConfig) { }
    createFilter() {
        return new RouterFilter();
    }
}
const ROUTER_FILTER_URL = 'type.googleapis.com/envoy.extensions.filters.http.router.v3.Router';
function parseConfig(encodedConfig) {
    return {
        typeUrl: ROUTER_FILTER_URL,
        config: null
    };
}
function setup() {
    (0, http_filter_1.registerHttpFilter)(ROUTER_FILTER_URL, {
        parseTopLevelFilterConfig: parseConfig,
        parseOverrideFilterConfig: parseConfig,
        httpFilterConstructor: RouterFilterFactory
    });
}
exports.setup = setup;
