// @grpc/grpc-js-xds/build/src/generated/envoy/admin/v3/ClientResourceStatus.js
"use strict";
// Original file: deps/envoy-api/envoy/admin/v3/config_dump_shared.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientResourceStatus = void 0;
/**
 * Resource status from the view of a xDS client, which tells the synchronization
 * status between the xDS client and the xDS server.
 */
var ClientResourceStatus;
(function (ClientResourceStatus) {
    /**
     * Resource status is not available/unknown.
     */
    ClientResourceStatus[ClientResourceStatus["UNKNOWN"] = 0] = "UNKNOWN";
    /**
     * Client requested this resource but hasn't received any update from management
     * server. The client will not fail requests, but will queue them until update
     * arrives or the client times out waiting for the resource.
     */
    ClientResourceStatus[ClientResourceStatus["REQUESTED"] = 1] = "REQUESTED";
    /**
     * This resource has been requested by the client but has either not been
     * delivered by the server or was previously delivered by the server and then
     * subsequently removed from resources provided by the server. For more
     * information, please refer to the :ref:`"Knowing When a Requested Resource
     * Does Not Exist" <xds_protocol_resource_not_existed>` section.
     */
    ClientResourceStatus[ClientResourceStatus["DOES_NOT_EXIST"] = 2] = "DOES_NOT_EXIST";
    /**
     * Client received this resource and replied with ACK.
     */
    ClientResourceStatus[ClientResourceStatus["ACKED"] = 3] = "ACKED";
    /**
     * Client received this resource and replied with NACK.
     */
    ClientResourceStatus[ClientResourceStatus["NACKED"] = 4] = "NACKED";
})(ClientResourceStatus = exports.ClientResourceStatus || (exports.ClientResourceStatus = {}));
