// @grpc/grpc-js-xds/build/src/fraction.js
"use strict";
/*
 * Copyright 2021 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.envoyFractionToFraction = exports.fractionToString = void 0;
function fractionToString(fraction) {
    return `${fraction.numerator}/${fraction.denominator}`;
}
exports.fractionToString = fractionToString;
const RUNTIME_FRACTION_DENOMINATOR_VALUES = {
    HUNDRED: 100,
    TEN_THOUSAND: 10000,
    MILLION: 1000000
};
function envoyFractionToFraction(envoyFraction) {
    return {
        numerator: envoyFraction.numerator,
        denominator: RUNTIME_FRACTION_DENOMINATOR_VALUES[envoyFraction.denominator]
    };
}
exports.envoyFractionToFraction = envoyFractionToFraction;
