// @grpc/grpc-js-xds/build/src/generated/envoy/service/status/v3/ConfigStatus.js
"use strict";
// Original file: deps/envoy-api/envoy/service/status/v3/csds.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigStatus = void 0;
/**
 * Status of a config from a management server view.
 */
var ConfigStatus;
(function (ConfigStatus) {
    /**
     * Status info is not available/unknown.
     */
    ConfigStatus[ConfigStatus["UNKNOWN"] = 0] = "UNKNOWN";
    /**
     * Management server has sent the config to client and received ACK.
     */
    ConfigStatus[ConfigStatus["SYNCED"] = 1] = "SYNCED";
    /**
     * Config is not sent.
     */
    ConfigStatus[ConfigStatus["NOT_SENT"] = 2] = "NOT_SENT";
    /**
     * Management server has sent the config to client but hasn’t received
     * ACK/NACK.
     */
    ConfigStatus[ConfigStatus["STALE"] = 3] = "STALE";
    /**
     * Management server has sent the config to client but received NACK. The
     * attached config dump will be the latest config (the rejected one), since
     * it is the persisted version in the management server.
     */
    ConfigStatus[ConfigStatus["ERROR"] = 4] = "ERROR";
})(ConfigStatus = exports.ConfigStatus || (exports.ConfigStatus = {}));
